### PR TITLE
Prometheus: Handle errors and warnings in buffered client

### LINKED ISSUE #48384: incorrect status code returned by Grafana when prom data source returns throttling error (429)
**What happened**:

When Prometheus data source returns a throttling response (status code: 429) for a query, it is converted to a 400 status code by Grafana server. Also, the error message returned by data source is eaten up by Grafana server. This bug is applicable for **_/query_** and _**/query_range**_ APIs. Rest all appears to be working fine.

**Behavior with /series, /labels, /label/{name}/values APIs (correct behavior):**

**Http status code:**
status returned by Grafana server to client is 429 (retaining what data source sent)

**Response message:**
error message (explaining why data source throttled) is also retained by Grafana server. i.e., user gets to know exactly why the query was throttled.

**The call to Grafana server by Grafana client (browser):**
http://localhost:3000/api/datasources/proxy/2/api/v1/series
http://localhost:3000/api/datasources/proxy/2/api/v1/labels
http://localhost:3000/api/datasources/proxy/2/api/v1/label/__name__/values

**See the snapshots:**
image

 

**Network tab:** 
image

image

**Behavior with /query, /query_range APIs (incorrect behavior):**

**Http status code:**
status returned by Grafana server to client is 400 in (see network tab below in snapshot – Grafana doesn’t retain what data source sent)
**Response message:** 
error message (explaining why data source throttled) is not retained by Grafana server. i.e., user doesn’t get to know exactly why the query was throttled.

**The call to Grafana server by Grafana client (browser) for both _/query_ and _/query_range_:**
http://localhost:3000/api/ds/query

**See the snapshots:**

**What you expected to happen**:

Expected behavior for Prometheus /query, /query_range APIs: 
- Correct error message to be returned to user
- Correct http response code to be returned to user

**How to reproduce it (as minimally and precisely as possible)**:
Setup Grafana locally and simulate a throttling response to be returned by Prometheus data source. Http Status code: 429 and some message in response body.

**Environment**:
- Grafana version: 8.5.0
- Data source type & version: Prometheus
- OS Grafana is installed on: Windows
- User OS & Browser: Windows

### PR DESCRIPTION
**Pull request by itsmylife**
Handle/Send the errors that are coming from Prometheus to the client.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/48384

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by marefr** [comment]
Nice. Is this ready for review? It's in draft. Curious, where you able to come up with a way to test this properly?

File: pkg/tsdb/prometheus/buffered/time_series_query.go [hunk 1]
```diff
@@ -155,6 +156,12 @@ func (b *Buffered) runQuery(ctx context.Context, query *PrometheusQuery) (backen
 	if query.RangeQuery {
 		rangeResponse, _, err := b.client.QueryRange(ctx, query.Expr, timeRange)
 		if err != nil {
+			var promErr *apiv1.Error
```
Comment (toddtreece) [inline pkg/tsdb/prometheus/buffered/time_series_query.go:159]: @kylebrandt recently changed how instant and range queries collect errors for the new client in https://github.com/grafana/grafana/pull/57698. should something similar be done here?

File: pkg/tsdb/prometheus/buffered/time_series_query.go [hunk 2]
```diff
@@ -175,6 +189,12 @@ func (b *Buffered) runQuery(ctx context.Context, query *PrometheusQuery) (backen
 	if query.ExemplarQuery {
 		exemplarResponse, err := b.client.QueryExemplars(ctx, query.Expr, timeRange.Start, timeRange.End)
 		if err != nil {
+			var promErr *apiv1.Error
+			if errors.As(err, &promErr) {
+				logger.Error("Exemplar query failed", "query", query.Expr, "error", err, "detail", promErr.Detail)
+				return backend.DataResponse{Error: fmt.Errorf("%w: details: %s", err, promErr.Detail)}, nil
```
[the remark this answers is not in the document]
Comment (itsmylife) [reply pkg/tsdb/prometheus/buffered/time_series_query.go:195]: You are right. We do not need this. I will revert.

File: pkg/tsdb/prometheus/buffered/time_series_query.go (hunk 1 above)
Comment (itsmylife) [reply pkg/tsdb/prometheus/buffered/time_series_query.go:159]: We will eventually deprecate buffered client so I do not want to spend too much time on this. But I will check what I can do. Thanks for the suggestion.

**Comment by itsmylife** [comment]
> Nice. Is this ready for review? It's in draft. Curious, where you able to come up with a way to test this properly?

It is still a draft. I need to provide tests before marking it ready. 
Also just because of the consistency I handled the warnings like it is handled here https://github.com/grafana/grafana/blob/main/pkg/util/converter/prom.go#L63-L70
But @toddtreece told me it is ok to not add warnings in each and every single frame. If I add the warning into the first frame it will be ok. Maybe I will handle this in a separate PR but want to share here for visibility.

**Review by kylebrandt — APPROVED** [review]

**Review by marefr — APPROVED** [review]
LGTM

As I suggested when we talked, backporting this to v9.2.x would be nice/my advice.

**Pull request merged**